### ISSUE TITLE
Instancing not working properly, and too many verts

### ISSUE THREAD

**Issue by kbrandwijk** [StjerneIdioten/I3D-Blender-Addon#39]
If I use 'duplicate linked' in Blender to create multiple instances of an object, the export result is not the same as in Giants exporter for 2.79 (or the NLD community exporter for 2.80). The translation is not applied succesfully, so all instances appear in the location of the original one.

![image](https://user-images.githubusercontent.com/852069/82515309-36da0680-9b18-11ea-8202-46e1c760ff67.png)

As you can see in the screenshot, the triangle count is 12 (the same as a single cube), as well as the vert count, but the cubes are overlapping.

![image](https://user-images.githubusercontent.com/852069/82515431-7f91bf80-9b18-11ea-92c8-aad2bb985bbf.png)

This is an export from the official Giants exporter. Tris 12, Verts 8, and the translateZ is in there like it should, showing the instanced second cube in the right position.

These two screenshots show something else. When I export a cube in 2.79, it is exported with 8 verts. If I export the same cube in 2.80 with your exporter, it uses 24 verts. On a big model, that becomes a problem quickly. Although I do not use sharp edges a lot for 'hero' parts of my model, they are used plenty in less visible parts, so that adds up.

Unfortunately, I use instancing a lot in the model I'm working on, and this means that I cannot use your exporter currently, so I hope there's an easy fix. I have attached both exports for comparison, if you need anything else, please let me know.

[i3d.zip](https://github.com/StjerneIdioten/I3D-Blender-Addon/files/4660271/i3d.zip)

**Comment by StjerneIdioten** [comment]
I cannot answer for the first part with the linked duplicates right now, but I do know that I have never actually tried that out, since I mainly use instancing to duplicate stuff. I am in the middle of a rework of the entire codebase (Which is almost done) so it make take a couple of weeks before I can try and resolve that other issue (If it isn't addresses as part of my rework)

For the part about my exporter having 24 verts vs the 8 verts of the giants exporter. 
I use the split normals data to export the normals, where as the giants exporter uses just the vertex normals. The difference is that what I export is the actual normal data for the vertices, where the Giants exporter always exports the averaged data per vertex. If I recall correctly, if you set the shading mode to smooth in blender, it should give you the same result for my exporter as for the Giants one, with the difference that mine actually gives more vertices still (I think this might be a numerical bug though, since it should be giving 8), but when you set the shading mode to "flat", which is what the default cube is at per default. Then my exporter correctly exports all of the vertex normals (Which if you look in blender and enable split vertex normals view in edit mode), where as GE still only exports it as if it was shaded smooth.
Essentially, when the cube is shaded flat in blender it should give 24 vertices (6 individual planes), when exported to Giants Engine, to keep the sharp edges. It is the same as adding the sharp modifier to an edge, which if I recall correctly, Giants Engine also ignores.

Btw the NLD community exporter and Giants 2.79 exporter are just about the same, when it comes to actually exporting things (NLD is just a fix that makes Giants version not crash in 2.80+)

And in the future it would be lovely if we could keep the issues to one bug per issue, just so it is easier to focus on a task and for documentation purposes :-)

I will make a note of this and look through the vertex normal export again, when I have finished reworking the code. Since that part is some of the first stuff written back when I knew just about nothing about blender and 3d graphics :-P

**Comment by kbrandwijk** [comment]
Good suggestion about the collection instancing, I've actually never used that, I guess old habits and all that. If that works fine, at least I'm no longer blocked on this.

I actually thought this was a single issue while writing it up, and that the high vert count was due to the incorrect instancing. Only at the very end did I notice by accident that the numbers stayed the same when I selected a single cube, and by then it was 4.15am and I didn't split it into 2 issues 😄 I'll make sure to do so next time though for tracking purposes, you are right.

Interesting to read how you are exporting normals. Does that mean that your exporter also support face weighted normals? That would be great. Like in the far right image in this infographic.

![image](https://user-images.githubusercontent.com/852069/82519461-0b5c1980-9b22-11ea-902c-f74ec1934350.png)

Either way, thanks again for your efforts!

**Comment by StjerneIdioten** [comment]
The "eww, hard edge" would be what you get, when you export the model with "shade flat" in blender. The "eww, ugly shading" is what you get, when you "shade smooth" in blender (And to my knowledge, it is also what the Giants exporter gives you no matter what you do) and I believe you can get the "beautiful" by using the weighted normals modifier in blender.

To the best of my knowledge, no matter how you mess with the normals in blender, the normal data should be stored in the place where I fetch them from when exporting. What I do is look through that normal data and for each unique normal you will also get a vertex since every vertex only has one normal. 

If I recall correctly I had someone draw up an example of different shading modes and I there was some face weighting or maybe it was something with painting the vertex weights, none the less you could mess around with the normals however you wished :-P

**Comment by StjerneIdioten** [comment]
@kbrandwijk Could I get you to test it again with the newest version of the addon? I just did a little test with two linked cubes in blender and it came out fine in GE.

![image](https://user-images.githubusercontent.com/11218511/84576936-a9a56d00-adb8-11ea-9ba6-73e45c9eab14.png)

![image](https://user-images.githubusercontent.com/11218511/84576942-af9b4e00-adb8-11ea-9515-510fe596de66.png)

**Comment by StjerneIdioten** [comment]
I've also looked into the whole vertex count thing and I figured out why my exporter gives 14 instead of 8 vertices for a standard cube with soft shading. Turns out that Giants exporter doesn't differentiate based on vertex UV's. So even though the cube only has 8 vertices it will have more unique ones (actually 14) if you consider the UV's.

![image](https://user-images.githubusercontent.com/11218511/84577584-f475b380-adbd-11ea-908a-effe6e883939.png)

I guess if you don't use the UV's for anything it will give you more vertices, but that should really not come up often and not in a way that would affect the vert count substantially.

**Comment by StjerneIdioten** [comment]
I will close the issue for now, since it seems that nothing is actually wrong at this point.

**Comment by kbrandwijk** [comment]
Sure, I'll retest the instancing somewhere over the next few days. Yeah that vertex count thing is probably fine. Like I said earlier, for me personally it's an edge case either way, because I always use bevel + face weighted normals.

**Comment by kbrandwijk** [comment]
@StjerneIdioten Did you ever revisit this normals export? Because it seems that, at least in the latest version, when I bevel + weighted normal by myself in Blender (having full control over the normals), the exporter still seems to think it needs to split up my vertices and look at the (artificial) 'split normals' that Blender comes up with. A simple beveled cube with 24 verts (verts that _do not_ need to be split up further), ends up with 48 (or weirdly enough with 77 on my machine) verts in GE, which is at this point, simply wrong.

You are looking at split normals, and generating vertices for each, while I believe you should not touch the verts at all (or at least make that an option) and only take the normal it already has in Blender (the actual, real normal, that is, not the split one). I spend a lot of time getting normals right (to the point of transplanting normals data when I cut a hole in something for example), and right now, all that effort is completely wasted because it doesn't get exported like I define it.

**Comment by ssnd292** [comment]
GE counts vertex normals plus UV split vertices, not just vertices.
"Split normals" arent just a "Blender Thing" - they are quite literally vertex normals. If you have a split face on an edge to create hard smoothing, the GPU duplicates the vertices for each face to singularly calculate the normals for that face. That also means having the same vertex three or for times.

See my first picture. It is a basic model with just average smoothing. 12 Vertices shown in both.
![2022-01-05 19_42_51-](https://user-images.githubusercontent.com/3914410/148271763-e0d28445-a4e7-4d20-82e1-a6aa170c17c6.jpg)

Now, the second picture I added some hard edtges in some places. If you now count the vertex normals you will see that it will actually count 14 - same as GE does.

![2022-01-05 19_45_09-](https://user-images.githubusercontent.com/3914410/148271850-aa1ddcb6-64a4-4371-9ff5-3ce2f2e5b2a4.jpg)

The extra vertices with an unwrapped model are coming from the UV vertices. So if you split your models UV you will have a higher vertex count. Still not quite sure why in this picture its only three more instead of 4 (since you have 4 extra unique vertices on the UV).

![2022-01-05 19_53_01-](https://user-images.githubusercontent.com/3914410/148272630-abcbe838-3cc4-4e2c-8a6a-0e0413659320.jpg)

This picture is just to show that its 4 extra vertices. Full Average Smoothing, just a UV Split added

![2022-01-05 19_52_25-](https://user-images.githubusercontent.com/3914410/148272712-46ee6423-3664-4e34-a84a-ea408d633979.jpg)

**Comment by ssnd292** [comment]
Regarding Normal orientation:
![2022-01-05 20_02_49-Blender](https://user-images.githubusercontent.com/3914410/148273909-5301e23e-57fc-47b5-8cd4-b6c32dc0ec34.jpg)
![2022-01-05 20_02_38-GIANTS Editor 64bit 9 0 2 - C__Users_ssnd2_Desktop_untitled i3d](https://user-images.githubusercontent.com/3914410/148273910-105138e3-fd7e-4934-aec1-8c48bdc2b9ec.jpg)
![2022-01-05 20_02_15-Blender](https://user-images.githubusercontent.com/3914410/148273912-0bc289ee-c719-48f7-9017-780a730f25ae.jpg)
